### PR TITLE
Dart: update benchmark to Dart 3.0

### LINKED ISSUE #146: Re-enable Dart benchmark
Dart benchmark stopped working. It was disabled temporarily, would be nice to re-enable it once the issue is identified and fixed.

### PR DESCRIPTION
**Pull request by amka**
**Summary of changes**
Changes introduced in this pull request:
- Update Dart SDK to 3.0
- Remove gRPC package deprecations
- Fix Docker build stage

**Reference issue to close (if applicable)**
Closes #146 

As fart as I think Dart is required for this benchmark 'cause it's one of the "official" grpc languages I fixed the benchmark. Current version of Dart is 3.0.5 so the SDK was updated to this release. Also dependencies and the pubspec file were updated accordingly.
I also renamed benchmark folder to be able to activate this benchmark. Testes locally and it worked like a charm.

Hope it help! Any commend are appreciated!

### CODE REVIEW

File: dart_grpc_bench/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+COPY dart_grpc_bench/pubspec.yaml /app/pubspec.yaml
+COPY proto /app/proto
+
+RUN dart pub get
+COPY dart_grpc_bench /app
+# Ensure packages are still up-to-date if anything has changed
+RUN dart pub get --offline
+RUN apt update && apt install -y protobuf-compiler
+RUN dart pub global activate protoc_plugin
+RUN mkdir -p lib/src/generated
```
Comment (He-Pin) [inline dart_grpc_bench/Dockerfile:13]: fold into one command?

Comment (amka) [reply dart_grpc_bench/Dockerfile]: > fold these `RUN` into one command?

Make sense. Done!

**Comment by LesnyRumcajs** [comment]
@amka Thanks! Could you please update the CI? You can either call `build.sh` (or `bench.sh`) with the Dart suite or run `./generate_ci.sh >.github/workflows/build.yml` and commit that.

**Comment by amka** [comment]
Sure, I will do it!

**Comment by LesnyRumcajs** [comment]
I'm a bit pushed for time; I'll try to check this out this weekend @amka!

**Comment by amka** [comment]
> I'm a bit pushed for time; I'll try to check this out this weekend @amka!

Oh, no worries! It's your repo! I'm just the guy who's trying to help :)

**Comment by LesnyRumcajs** [comment]
Not sure why the script generated it for you this way, let's not bother with it. If you'd put `dart` after `d` to make CI happy it should be fine. Oh, and if you could please do these changes with `git mv` that'd be great, the old `dart_grpc_onhold` is still lingering there.

**Comment by amka** [comment]
> Not sure why the script generated it for you this way, let's not bother with it. If you'd put `dart` after `d` to make CI happy it should be fine. Oh, and if you could please do these changes with `git mv` that'd be great, the old `dart_grpc_onhold` is still lingering there.

@LesnyRumcajs  Not sure I understand you correctly about git mv. Do you mean to replace the old one `_onhold` with a new `_bench` folder?

**Comment by LesnyRumcajs** [comment]
> > Not sure why the script generated it for you this way, let's not bother with it. If you'd put `dart` after `d` to make CI happy it should be fine. Oh, and if you could please do these changes with `git mv` that'd be great, the old `dart_grpc_onhold` is still lingering there.
> 
> @LesnyRumcajs Not sure I understand you correctly about git mv. Do you mean to replace the old one `_onhold` with a new `_bench` folder?

Yes, precisely. The `_onhold` suffix means that the particular bench suite was disabled for whatever reason. Your contribution effectively replaces it. At the current state of the PR, you can just remove it.

**Comment by LesnyRumcajs** [comment]
@amka did you try running it locally recently? It seems to be failing in the CI.

**Comment by amka** [comment]
I've got a weird issue. When I ran `sh ./bench.sh dart_grpc_bench` locally for the first time it worked like a charm. When I rerun it - it fails quickly. We'll investigate it.

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -239,6 +239,42 @@ jobs:
       env:
         GRPC_IMAGE_NAME: ${{ needs.set-image-name.outputs.name }}
 
+
```
Comment (LesnyRumcajs) [inline .github/workflows/build.yml:242]:
```suggestion
```

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -239,6 +239,41 @@ jobs:
       env:
         GRPC_IMAGE_NAME: ${{ needs.set-image-name.outputs.name }}
 
+  dart_grpc_bench:
```
Comment (LesnyRumcajs) [inline .github/workflows/build.yml:242]:
```suggestion

  dart_grpc_bench:
```

**Comment by LesnyRumcajs** [comment]
@amka I fixed the meta check (whitespace issue), now the dart server is failing.  I debugged it a bit, locally. The server seems to be crashing during the warmup period, from time to time. See the error message:
```
❯ docker logs dart_grpc_bench -f
Server listening on port 50051...
Unhandled exception:
Null check operator used on a null value
#0      ServerHandler.sendTrailers (package:grpc/src/server/handler.dart:384)
#1      ServerHandler._sendError (package:grpc/src/server/handler.dart:459)
#2      ServerHandler._onResponse (package:grpc/src/server/handler.dart:327)
#3      _RootZone.runUnaryGuarded (dart:async/zone.dart:1594)
#4      _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:339)
#5      _BufferingStreamSubscription._add (dart:async/stream_impl.dart:271)
#6      _SyncStreamControllerDispatch._sendData (dart:async/stream_controller.dart:784)
#7      _StreamController._add (dart:async/stream_controller.dart:658)
#8      new Stream.fromFuture.<anonymous closure> (dart:async/stream.dart:250)
#9      Future._propagateToListeners.handleValueCallback (dart:async/future_impl.dart:846)
#10     Future._propagateToListeners (dart:async/future_impl.dart:875)
#11     Future._chainCoreFutureSync (dart:async/future_impl.dart:575)
#12     Future._chainCoreFutureAsync.<anonymous closure> (dart:async/future_impl.dart:613)
#13     _microtaskLoop (dart:async/schedule_microtask.dart:40)
#14     _startMicrotaskLoop (dart:async/schedule_microtask.dart:49)
#15     _runPendingImmediateCallback (dart:isolate-patch/isolate_patch.dart:118)
#16     _RawReceivePort._handleMessage (dart:isolate-patch/isolate_patch.dart:185)
```

It might make sense to bump the dependencies, perhaps the ones defined in `pubspec.yaml` and `pubspec.lock` are faulty?

**Comment by LesnyRumcajs** [comment]
Stale. Closing.

**Pull request closed without merging**